src/main/java/com/google/devtools/build/lib/hash/Blake3Hasher.java

```
@@ -0,0 +1,115 @@
+  // written, a single JNI call will be made that initializes, hashes, and
+  // cleans up the hasher, rather than making separate calls for each operation.
+  public static final int ONESHOT_THRESHOLD = 8 * 1024;
+
+  private final ReentrantLock rl = new ReentrantLock();
+  private ByteBuffer buffer;
+  private long hasher = -1;
+
+  public boolean isAllocated() {
+    return (hasher != -1);
+  }
+
+  private void initOnce() {
+    if (!isAllocated()) {
+      hasher = Blake3JNI.allocate_and_initialize_hasher();
+    }
+  }
+
+  public Blake3Hasher() {}
+
+  private void resetBuffer(int minLength) {
+    int length = Math.max(ONESHOT_THRESHOLD, minLength);
+    if (buffer == null || buffer.capacity() < length) {
+      buffer = ByteBuffer.allocateDirect(length);
+      buffer.order(ByteOrder.nativeOrder());
+    }
+    buffer.clear();
+  }
+
+  public void update(byte[] data, int offset, int length) {
+    rl.lock();
+
+    if (buffer == null) {
+      resetBuffer(length);
+    }
+
+    try {
+      if (buffer.remaining() < length) {
+        initOnce();
+        Blake3JNI.blake3_hasher_update(hasher, buffer, 0, buffer.position());
+        resetBuffer(length);
+      }
+      buffer.put(data, offset, length);
+    } finally {
+      rl.unlock();
+    }
+  }
+
+  public void update(byte[] data) {
+    update(data, 0, data.length);
+  }
+
+  public byte[] getOutput(int outputLength) throws IllegalArgumentException {
+    rl.lock();
+    byte[] retByteArray = new byte[outputLength];
+
+    try {
+      if (!isAllocated() && buffer != null) {
+        Blake3JNI.blake3_hasher_oneshot(
+            hasher, buffer, buffer.position(), retByteArray, outputLength);
+      } else {
+        initOnce();
+        if (buffer != null && buffer.position() > 0) {
+          Blake3JNI.blake3_hasher_update(hasher, buffer, 0, buffer.position());
+        }
+        Blake3JNI.blake3_hasher_finalize_and_close(hasher, retByteArray, outputLength);
+        hasher = -1;
```

in addition to cleaning up the hasher, you may want to imitate how the other guava hashers set a "done" bit in hash()--it shouldn't matter, but they fail fast if you somehow coded things up to call hash() twice.

---

Done.